### PR TITLE
feat: Support OPTIONAL MATCH clause

### PR DESCRIPTION
**Pull request by protodef**

### CODE REVIEW

File: src/backend/parser/parse_graph.c [hunk 1]
```diff
@@ -278,43 +290,53 @@ transformCypherMatchClause(ParseState *pstate, CypherClause *clause)
...
 	Query	   *qry;
 	RangeTblEntry *rte;
-	Node	   *qual;
+	Node	   *qual = NULL;
 
 	qry = makeNode(Query);
 	qry->commandType = CMD_SELECT;
 
-	if (detail->where != NULL)
+	if (detail->optional && clause->prev != NULL)
```
Comment (htlim) [inline src/backend/parser/parse_graph.c]: If an optional match is top of statement, is this proceeded like normal match clause?

If using optional match and no matches are found, does the right to return a null value?

Comment (htlim) [inline src/backend/parser/parse_graph.c]: optional match (a {'not':'exist'}) return a;

we return 0 rows, is this right?
neo4j returns a row with null value.

File: src/backend/parser/parse_graph.c [hunk 2]
```diff
@@ -329,6 +351,160 @@ transformCypherMatchClause(ParseState *pstate, CypherClause *clause)
...
+	/* transform LEFT */
+	l_alias = makeAliasNoDup(CYPHER_OPTLEFT_ALIAS, NIL);
+	j->larg = transformClauseJoin(pstate, clause->prev, l_alias, &l_rte,
+								  &l_nsitem);
+
+	/* temporal namespace for RIGHT */
+	Assert(pstate->p_namespace == NIL);
+	pstate->p_namespace = lappend(pstate->p_namespace, l_nsitem);
+
+	/* Transform RIGHT. Prevent `claus`e from being transformed infinitely. */
```
Comment (htlim) [inline src/backend/parser/parse_graph.c]: typo?

Comment (protodef) [inline src/backend/parser/parse_graph.c]: Fixed :)

File: src/backend/parser/parse_graph.c (hunk 1 above)
Comment (protodef) [inline src/backend/parser/parse_graph.c]: Yes. It is intentional behavior.

**Pull request merged**